### PR TITLE
PYTHON-2769 Test redaction of replies to security-sensitive commands

### PR DESCRIPTION
**Pull request by ShaneHarvey**

### CODE REVIEW

File: pymongo/monitoring.py [hunk 1]
```diff
@@ -512,7 +513,7 @@ def register(listener):
 # The "hello" command is also deemed sensitive when attempting speculative
 # authentication.
 def _is_speculative_authenticate(command_name, doc):
-    if (command_name.lower() in ('hello', 'ismaster') and
+    if (command_name.lower() in (HelloCompat.CMD, HelloCompat.LEGACY_CMD) and
```
Comment (ShaneHarvey) [inline pymongo/monitoring.py:516]: Unrelated but I figured I'd change this while I was looking at this code.

File: pymongo/monitoring.py [hunk 2]
```diff
@@ -1362,9 +1364,14 @@ def publish_command_success(self, duration, reply, command_name,
...
           - `op_id`: The (optional) operation id for this operation.
           - `service_id`: The service_id this command was sent to, or ``None``.
+          - `speculative_hello`: Was the command sent with speculative auth?
         """
         if op_id is None:
             op_id = request_id
+        if speculative_hello:
+            # Redact entire response when the command started contained
+            # speculativeAuthenticate.
+            reply = {}
```
Comment (ShaneHarvey) [inline pymongo/monitoring.py:1374]: To understand why we need to do this see this new test: https://github.com/mongodb/specifications/blob/289f405/source/command-monitoring/tests/unified/redacted-commands.yml#L232-L242

File: test/unified_format.py [hunk 3]
```diff
@@ -409,8 +409,8 @@ def _operation_type(self, spec, actual, key_to_compare):
                 t for alias in spec for t in self.__type_alias_to_type(alias)])
         else:
             permissible_types = self.__type_alias_to_type(spec)
-        self.test.assertIsInstance(
-            actual[key_to_compare], permissible_types)
+        value = actual[key_to_compare] if key_to_compare else actual
+        self.test.assertIsInstance(value, permissible_types)
```
Comment (ShaneHarvey) [inline test/unified_format.py:413]: We can assert the type of a result at the top-level in which case the key_to_compare is None.

File: test/unified_format.py [hunk 4]
```diff
@@ -822,6 +822,8 @@ def _collectionOperation_find(self, target, *args, **kwargs):
 
     def _collectionOperation_createFindCursor(self, target, *args, **kwargs):
         self.__raise_if_unsupported('find', target, Collection)
+        if 'filter' not in kwargs:
+            self.fail('createFindCursor requires a "filter" argument')
```
Comment (ShaneHarvey) [inline test/unified_format.py:826]: See https://github.com/mongodb/specifications/blob/ed4e62b/source/unified-test-format/tests/valid-fail/entity-find-cursor.yml#L23-L27

File: test/unified_format.py [hunk 5]
```diff
@@ -909,7 +911,9 @@ def run_entity_operation(self, spec):
         try:
             result = cmd(**dict(arguments))
         except Exception as exc:
-            if ignore:
+            # Ignore all operation errors but to avoid masking bugs don't
+            # ignore things like TypeError and ValueError.
+            if ignore and isinstance(exc, (PyMongoError,)):
```
Comment (ShaneHarvey) [inline test/unified_format.py:916]: See https://github.com/mongodb/specifications/blob/ed4e62b48939378b814cbd7d9d4939a2c1076a6b/source/unified-test-format/tests/valid-fail/ignoreResultAndError.yml#L37-L43

**Comment by ShaneHarvey** [comment]
I need to fix the following failures:
```
 [2021/07/13 20:21:36.016] FAIL [0.060s]: test_legacy_bulkWrite_A_successful_mixed_bulk_write (test_command_monitoring_legacy.TestAllScenarios)
 [2021/07/13 20:21:36.016] ----------------------------------------------------------------------
 [2021/07/13 20:21:36.016] Traceback (most recent call last):
 [2021/07/13 20:21:36.016]   File "/data/mci/71ac8046c2c5a7b96286b05213264b86/src/test/test_command_monitoring_legacy.py", line 200, in run_scenario
 [2021/07/13 20:21:36.016]     self.assertEqual(val, actual[key])
 [2021/07/13 20:21:36.016] AssertionError: Lists differ: [{'q': {'_id': 3}, 'u': {'$set': {'x': 333}}}] != [SON([('q', {'_id': 3}), ('u', {'$set': {'[45 chars]e)])]
 [2021/07/13 20:21:36.016] First differing element 0:
 [2021/07/13 20:21:36.016] {'q': {'_id': 3}, 'u': {'$set': {'x': 333}}}
 [2021/07/13 20:21:36.016] SON([('q', {'_id': 3}), ('u', {'$set': {'[44 chars]se)])
 [2021/07/13 20:21:36.016] - [{'q': {'_id': 3}, 'u': {'$set': {'x': 333}}}]
 [2021/07/13 20:21:36.016] + [{'multi': False,
 [2021/07/13 20:21:36.016] +   'q': {'_id': 3},
 [2021/07/13 20:21:36.016] +   'u': {'$set': {'x': 333}},
 [2021/07/13 20:21:36.016] +   'upsert': False}]
 [2021/07/13 20:21:36.016] ======================================================================
 [2021/07/13 20:21:36.016] FAIL [0.009s]: test_legacy_updateMany_A_successful_update_many (test_command_monitoring_legacy.TestAllScenarios)
 [2021/07/13 20:21:36.016] ----------------------------------------------------------------------
 [2021/07/13 20:21:36.016] Traceback (most recent call last):
 [2021/07/13 20:21:36.016]   File "/data/mci/71ac8046c2c5a7b96286b05213264b86/src/test/test_command_monitoring_legacy.py", line 200, in run_scenario
 [2021/07/13 20:21:36.016]     self.assertEqual(val, actual[key])
 [2021/07/13 20:21:36.016] AssertionError: Lists differ: [{'q': {'_id': {'$gt': 1}}, 'u': {'$inc': [21 chars]rue}] != [SON([('q', {'_id': {'$gt': 1}}), ('u', {'[51 chars]e)])]
 [2021/07/13 20:21:36.016] First differing element 0:
 [2021/07/13 20:21:36.016] {'q': {'_id': {'$gt': 1}}, 'u': {'$inc': [20 chars]True}
 [2021/07/13 20:21:36.016] SON([('q', {'_id': {'$gt': 1}}), ('u', {'[50 chars]se)])
 [2021/07/13 20:21:36.016] - [{'multi': True, 'q': {'_id': {'$gt': 1}}, 'u': {'$inc': {'x': 1}}}]
 [2021/07/13 20:21:36.016] + [{'multi': True,
 [2021/07/13 20:21:36.016] +   'q': {'_id': {'$gt': 1}},
 [2021/07/13 20:21:36.016] +   'u': {'$inc': {'x': 1}},
 [2021/07/13 20:21:36.016] +   'upsert': False}]
 [2021/07/13 20:21:36.016] ======================================================================
 [2021/07/13 20:21:36.016] FAIL [0.009s]: test_legacy_updateMany_A_successful_update_many_command_with_write_errors (test_command_monitoring_legacy.TestAllScenarios)
 [2021/07/13 20:21:36.016] ----------------------------------------------------------------------
 [2021/07/13 20:21:36.016] Traceback (most recent call last):
 [2021/07/13 20:21:36.016]   File "/data/mci/71ac8046c2c5a7b96286b05213264b86/src/test/test_command_monitoring_legacy.py", line 200, in run_scenario
 [2021/07/13 20:21:36.016]     self.assertEqual(val, actual[key])
 [2021/07/13 20:21:36.016] AssertionError: Lists differ: [{'q': {'_id': {'$gt': 1}}, 'u': {'$nothin[25 chars]rue}] != [SON([('q', {'_id': {'$gt': 1}}), ('u', {'[55 chars]e)])]
 [2021/07/13 20:21:36.016] First differing element 0:
 [2021/07/13 20:21:36.016] {'q': {'_id': {'$gt': 1}}, 'u': {'$nothin[24 chars]True}
 [2021/07/13 20:21:36.017] SON([('q', {'_id': {'$gt': 1}}), ('u', {'[54 chars]se)])
 [2021/07/13 20:21:36.017] - [{'multi': True, 'q': {'_id': {'$gt': 1}}, 'u': {'$nothing': {'x': 1}}}]
 [2021/07/13 20:21:36.017] + [{'multi': True,
 [2021/07/13 20:21:36.017] +   'q': {'_id': {'$gt': 1}},
 [2021/07/13 20:21:36.017] +   'u': {'$nothing': {'x': 1}},
 [2021/07/13 20:21:36.017] +   'upsert': False}]
 [2021/07/13 20:21:36.017] ======================================================================
 [2021/07/13 20:21:36.017] FAIL [0.009s]: test_legacy_updateOne_A_successful_update_one (test_command_monitoring_legacy.TestAllScenarios)
 [2021/07/13 20:21:36.017] ----------------------------------------------------------------------
 [2021/07/13 20:21:36.017] Traceback (most recent call last):
 [2021/07/13 20:21:36.017]   File "/data/mci/71ac8046c2c5a7b96286b05213264b86/src/test/test_command_monitoring_legacy.py", line 200, in run_scenario
 [2021/07/13 20:21:36.017]     self.assertEqual(val, actual[key])
 [2021/07/13 20:21:36.017] AssertionError: Lists differ: [{'q': {'_id': {'$gt': 1}}, 'u': {'$inc': {'x': 1}}}] != [SON([('q', {'_id': {'$gt': 1}}), ('u', {'[52 chars]e)])]
 [2021/07/13 20:21:36.017] First differing element 0:
 [2021/07/13 20:21:36.017] {'q': {'_id': {'$gt': 1}}, 'u': {'$inc': {'x': 1}}}
 [2021/07/13 20:21:36.017] SON([('q', {'_id': {'$gt': 1}}), ('u', {'[51 chars]se)])
 [2021/07/13 20:21:36.017] - [{'q': {'_id': {'$gt': 1}}, 'u': {'$inc': {'x': 1}}}]
 [2021/07/13 20:21:36.017] + [{'multi': False,
 [2021/07/13 20:21:36.017] +   'q': {'_id': {'$gt': 1}},
 [2021/07/13 20:21:36.017] +   'u': {'$inc': {'x': 1}},
 [2021/07/13 20:21:36.017] +   'upsert': False}]
 [2021/07/13 20:21:36.017] ======================================================================
 [2021/07/13 20:21:36.017] FAIL [0.009s]: test_legacy_updateOne_A_successful_update_one_command_with_write_errors (test_command_monitoring_legacy.TestAllScenarios)
 [2021/07/13 20:21:36.017] ----------------------------------------------------------------------
 [2021/07/13 20:21:36.017] Traceback (most recent call last):
 [2021/07/13 20:21:36.017]   File "/data/mci/71ac8046c2c5a7b96286b05213264b86/src/test/test_command_monitoring_legacy.py", line 200, in run_scenario
 [2021/07/13 20:21:36.017]     self.assertEqual(val, actual[key])
 [2021/07/13 20:21:36.017] AssertionError: Lists differ: [{'q': {'_id': {'$gt': 1}}, 'u': {'$nothing': {'x': 1}}}] != [SON([('q', {'_id': {'$gt': 1}}), ('u', {'[56 chars]e)])]
 [2021/07/13 20:21:36.017] First differing element 0:
 [2021/07/13 20:21:36.017] {'q': {'_id': {'$gt': 1}}, 'u': {'$nothing': {'x': 1}}}
 [2021/07/13 20:21:36.017] SON([('q', {'_id': {'$gt': 1}}), ('u', {'[55 chars]se)])
 [2021/07/13 20:21:36.017] - [{'q': {'_id': {'$gt': 1}}, 'u': {'$nothing': {'x': 1}}}]
 [2021/07/13 20:21:36.017] + [{'multi': False,
 [2021/07/13 20:21:36.017] +   'q': {'_id': {'$gt': 1}},
 [2021/07/13 20:21:36.017] +   'u': {'$nothing': {'x': 1}},
 [2021/07/13 20:21:36.017] +   'upsert': False}]
 [2021/07/13 20:21:36.017] ======================================================================
 [2021/07/13 20:21:36.017] FAIL [0.009s]: test_legacy_updateOne_A_successful_update_one_with_upsert_when_the_upserted_id_is_not_an_object_id (test_command_monitoring_legacy.TestAllScenarios)
 [2021/07/13 20:21:36.017] ----------------------------------------------------------------------
 [2021/07/13 20:21:36.017] Traceback (most recent call last):
 [2021/07/13 20:21:36.017]   File "/data/mci/71ac8046c2c5a7b96286b05213264b86/src/test/test_command_monitoring_legacy.py", line 200, in run_scenario
 [2021/07/13 20:21:36.017]     self.assertEqual(val, actual[key])
 [2021/07/13 20:21:36.017] AssertionError: Lists differ: [{'q': {'_id': 4}, 'u': {'$inc': {'x': 1}}[13 chars]rue}] != [SON([('q', {'_id': 4}), ('u', {'$inc': {'[42 chars]e)])]
 [2021/07/13 20:21:36.017] First differing element 0:
 [2021/07/13 20:21:36.017] {'q': {'_id': 4}, 'u': {'$inc': {'x': 1}}, 'upsert': True}
 [2021/07/13 20:21:36.017] SON([('q', {'_id': 4}), ('u', {'$inc': {'[41 chars]ue)])
 [2021/07/13 20:21:36.017] - [{'q': {'_id': 4}, 'u': {'$inc': {'x': 1}}, 'upsert': True}]
 [2021/07/13 20:21:36.017] + [{'multi': False,
 [2021/07/13 20:21:36.017] +   'q': {'_id': 4},
 [2021/07/13 20:21:36.017] +   'u': {'$inc': {'x': 1}},
 [2021/07/13 20:21:36.017] +   'upsert': True}]
 [2021/07/13 20:21:36.017] ======================================================================
 [2021/07/13 20:21:36.017] FAIL [0.014s]: test_cursors_can_be_created,_iterated,_and_closed (test_unified_format.TestUnifiedTestFormatValidPassEntityFindCursor)
 [2021/07/13 20:21:36.017] ----------------------------------------------------------------------
 [2021/07/13 20:21:36.017] Traceback (most recent call last):
 [2021/07/13 20:21:36.017]   File "/data/mci/71ac8046c2c5a7b96286b05213264b86/src/test/unified_format.py", line 1127, in test_case
 [2021/07/13 20:21:36.017]     self.run_scenario(spec)
 [2021/07/13 20:21:36.017]   File "/data/mci/71ac8046c2c5a7b96286b05213264b86/src/test/unified_format.py", line 1114, in run_scenario
 [2021/07/13 20:21:36.017]     self.check_events(spec.get('expectEvents', []))
 [2021/07/13 20:21:36.017]   File "/data/mci/71ac8046c2c5a7b96286b05213264b86/src/test/unified_format.py", line 1066, in check_events
 [2021/07/13 20:21:36.017]     self.match_evaluator.match_event(
 [2021/07/13 20:21:36.017]   File "/data/mci/71ac8046c2c5a7b96286b05213264b86/src/test/unified_format.py", line 571, in match_event
 [2021/07/13 20:21:36.017]     self.match_result(reply, actual.reply)
 [2021/07/13 20:21:36.017]   File "/data/mci/71ac8046c2c5a7b96286b05213264b86/src/test/unified_format.py", line 513, in match_result
 [2021/07/13 20:21:36.017]     return self._match_document(
 [2021/07/13 20:21:36.017]   File "/data/mci/71ac8046c2c5a7b96286b05213264b86/src/test/unified_format.py", line 501, in _match_document
 [2021/07/13 20:21:36.017]     self.match_result(value, actual[key], in_recursive_call=True)
 [2021/07/13 20:21:36.017]   File "/data/mci/71ac8046c2c5a7b96286b05213264b86/src/test/unified_format.py", line 513, in match_result
 [2021/07/13 20:21:36.017]     return self._match_document(
 [2021/07/13 20:21:36.017]   File "/data/mci/71ac8046c2c5a7b96286b05213264b86/src/test/unified_format.py", line 497, in _match_document
 [2021/07/13 20:21:36.017]     if self._evaluate_if_special_operation(expectation, actual, key):
 [2021/07/13 20:21:36.017]   File "/data/mci/71ac8046c2c5a7b96286b05213264b86/src/test/unified_format.py", line 482, in _evaluate_if_special_operation
 [2021/07/13 20:21:36.017]     self._evaluate_special_operation(
 [2021/07/13 20:21:36.017]   File "/data/mci/71ac8046c2c5a7b96286b05213264b86/src/test/unified_format.py", line 447, in _evaluate_special_operation
 [2021/07/13 20:21:36.017]     method(spec, actual, key_to_compare)
 [2021/07/13 20:21:36.017]   File "/data/mci/71ac8046c2c5a7b96286b05213264b86/src/test/unified_format.py", line 413, in _operation_type
 [2021/07/13 20:21:36.017]     self.test.assertIsInstance(value, permissible_types)
 [2021/07/13 20:21:36.017] AssertionError: 5052123104391 is not an instance of (<class 'bson.int64.Int64'>,)
```

File: pymongo/monitoring.py (hunk 1 above)
Comment (ShaneHarvey) [reply pymongo/monitoring.py]: Backing this out to avoid merge conflicts.

File: pymongo/message.py [hunk 6]
```diff
@@ -1503,7 +1504,7 @@ class _OpReply(object):
 
     def __init__(self, flags, cursor_id, number_returned, documents):
         self.flags = flags
-        self.cursor_id = cursor_id
+        self.cursor_id = Int64(cursor_id)
```
Comment (ShaneHarvey) [inline pymongo/message.py:1507]: See https://jira.mongodb.org/browse/DRIVERS-1850.

When we decode a OP_REPLY cursor_id we decode it to a regular python int (with unlimited precision).
When we decode a command response the cursor.id field is decoded to a special bson.Int64 class.
So on <=3.0 the cursor id isn’t a “long”.

This change makes the cursor id always be Int64 regardless of the server version.

**Comment by ShaneHarvey** [comment]
This is ready for review.

File: pymongo/message.py (hunk 6 above)
Comment (prashantmital) [reply pymongo/message.py:1507]: Nice discovery.

**Review by prashantmital — APPROVED** [review]

**Pull request merged**